### PR TITLE
Create database package

### PR DESCRIPTION
**Pull request by arthur29**
WHY
======

Leaderboard how it currently is don't have a way to surround the database and your way to work. This PR proposes to create a `database` interface that will close database logic inside it to other packages interact. And a new Redis implementation that implements a database interface.

Another thing is that UnknownError is a bad name for errors that weren't handled by the package, so `redis.UnknownError` will be renamed to be `GeneralError`.

And finally, a mock implementation of `database` will be added to be used by tests.

WHAT WAS DONE
======
* Create `database` interface
* Move `leaderboard/redis` into `leaderboard/database`
* Create `database/mock`
* Change `UnknownError` to `GeneralError`

### CODE REVIEW

File: leaderboard/database/redis.go [hunk 1]
```diff
@@ -0,0 +1,209 @@
...
+
+	"github.com/topfreegames/podium/leaderboard/database/redis"
+)
+
+// Redis is a type that implements database interface with redis client
+type Redis struct {
+	redis.Redis
+}
+
+var _ Database = &Redis{}
```
Comment (mathnogueira) [inline leaderboard/database/redis.go:15]: Why are you creating an anonymous variable here?

Comment (arthur29) [reply leaderboard/database/redis.go:15]: This is the way to assert at compile time that a type implements a certain interface. In this case, I am asserting that Redis implements Database interface

Comment (mathnogueira) [reply leaderboard/database/redis.go:15]: Shouldn't this be done in a test file?

Comment (mathnogueira) [reply leaderboard/database/redis.go:15]: Very interesting btw

Comment (arthur29) [reply leaderboard/database/redis.go:15]: Hm, how other private types as `clusterClient` and `standaloneClient` inside `redis` package should do this same kind of assertion, and test is on "another" package . I prefer to keep all type assertions inside the same file.

Comment (arthur29) [reply leaderboard/database/redis.go:15]: https://github.com/topfreegames/podium/pull/45/files#diff-ba0a1ca105e821459b703bf496c548b2401d1ce3e0ca43f60cd0e449130d01fdR15

**Review by mathnogueira — APPROVED** [review]

**Pull request merged**